feat(ansible): make ansible verbosity configurable

djzager opened the pull request:
The primary objective of this change is to make the verbosity of
individual `ansible-runner` command invocations configurable (not just
"-vv"). In order to accomplish this:

* A new flag was added to the AnsibleOperatorFlags (ansible-verbosity)
  with a default value of 2.
* The entire flagset is given to the watches pkg Load() to provide one
  place for handling command line arguments, watches config values, and
  environment variables. This makes an individual watch more useful when
  creating a runner and adding a GVK to the controller. Also, this puts
  new fields on the Watch struct (MaxWorkers and AnsibleVerbosity).
* New functions were added to the watches pkg -- New(gvk) instantiates a
  Watch with sensible defaults. Validate() is used to verify a given
  watch meets certain criteria (legit path to playbook/role and proper
  finalizer). This makes it possible for someone who only knows their
  GVK and role/playbook to easily create a watch that can be used with
  the runner pkg.
* When creating a runner from a watch, we will now validate it using
  watches.Validate(). Essentially the watches pkg should know what is
  and isn't a valid watch.
* New helper functions in the runner pkg -- ansibleVerbosityString to
  convert an integer AnsibleVerbosity to a string (ie. 2 becomes "-vv"),
  playbookCmdFunc, and roleCmdFunc consolidate the logic for creating
  the cmdFund.

djzager commented on pkg/ansible/runner/runner_test.go:
```diff
@@ -26,32 +26,20 @@ import (
 	ident := "test"
 	inputDirPath := "/test/path"
-	var path string
+	maxArtifacts := 1
```
Since this is an argument to the `cmdFunc` it was silly to do an expected vs actual...just check that given the arguments `ident`, `inputDirPath`, and `maxArtifacts` that the `cmdFunc` on the created runner matches what we expect based on the particular watch under test.

djzager commented on pkg/ansible/runner/runner_test.go:
```diff
@@ -63,111 +51,88 @@ func checkCmdFunc(t *testing.T, cmdFunc cmdFuncType, role, playbook string, watc
+	if err != nil {
+		t.Fatalf("Unable to get working director: %v", err)
+	}
+	validPlaybook := filepath.Join(cwd, "testdata", "playbook.yml")
```
Since we are now validating watches, the role and playbook paths have to be legit.

djzager commented:
/test e2e-aws-ansible

djzager commented:
/test e2e-aws-ansible

djzager commented:
/retest

jmrodri commented:
/lgtm

camilamacedo86 commented:
HI @djzager, 

Really great work here 👍 . However, shows that it is missing update the docs regards these changes.. could you please add this part in the PR?

djzager commented:
@camilamacedo86 I'm looking through to see if there is any more documenting that I need to do.

camilamacedo86 commented on CHANGELOG.md:
```diff
@@ -8,6 +8,7 @@
 - Ansible based operators now gather and serve metrics about each custom resource on port 8686 of the metrics service. ([#1723](https://github.com/operator-framework/operator-sdk/pull/1723))
 - Added the Go version, OS, and architecture to the output of `operator-sdk version` ([#1863](https://github.com/operator-framework/operator-sdk/pull/1863))
 - Added support for `ppc64le-linux` for the `operator-sdk` binary and the Helm operator base image. ([#1533](https://github.com/operator-framework/operator-sdk/pull/1533))
+- Ansible based operators now support verbosity configuration via the `ansible-verbosity` flag at the command line. ([#1852](https://github.com/operator-framework/operator-sdk/pull/1852))
```
```suggestion
- Added support of verbosity configuration for Ansible-based Operators via the `ansible-verbosity` flag. ([#1852](https://github.com/operator-framework/operator-sdk/pull/1852))
```

camilamacedo86 replied:
Is not this change cable of to cause break changes in the current projects?

djzager replied:
The proposed changes should not break current projects or require any changes in those projects. If you have a scenario or code path in mind I should certainly fix it.

camilamacedo86 commented on pkg/ansible/run.go:
```diff
@@ -107,7 +105,7 @@ func Run(flags *aoflags.AnsibleOperatorFlags) error {
 			GVK:             w.GroupVersionKind,
 			Runner:          runner,
 			ManageStatus:    w.ManageStatus,
-			MaxWorkers:      getMaxWorkers(w.GroupVersionKind, flags.MaxWorkers),
```
So, currently, we can define the MaxWorkers by CRD. Why should we change it?

djzager replied:
The two places to set `MaxWorkers` are on the command line or via an environment variable. The only changes made here are to put the resolution of those values in the watches package + storing those values on the watches struct to be reused throughout execution.

camilamacedo86 commented on pkg/ansible/runner/runner.go:
```diff
@@ -52,56 +52,65 @@ type Runner interface {
 }
 
+func ansibleVerbosityString(verbosity int) string {
+	if verbosity > 0 {
```
Would not it can be as v, or vv, or vvv, or vvvv? See https://docs.ansible.com/ansible/latest/cli/ansible-playbook.html#cmdoption-ansible-playbook-v

So, wdyt about your solution allow the user to inform it?

djzager replied:
1. To me it made more sense to make this the integer value of the verbosity level. I don't believe this would be too surprising to Ansible users as that is how [`debug`](https://docs.ansible.com/ansible/latest/modules/debug_module.html#parameters) works. I also wanted to avoid any possible confusion with the operator-sdk's own verbosity.

camilamacedo86 replied:
But the `operator-sdk's own verbosity` is in integer. 

So, IMHO it should be done in the same way, which means we could have an env var to setup it  as follows:

```
env:
  - name: ANSIBLE_ VERBOSITY_LEVEL
     value: "vvvv" # it could be  v, or vv, or vvv, or vvvv
```

Also, IHMO we should keep the same values used by ansible instead of creating a new nomenclature and a from/to map for it.

camilamacedo86 commented on pkg/ansible/runner/runner_test.go:
```diff
@@ -63,111 +51,88 @@ func checkCmdFunc(t *testing.T, cmdFunc cmdFuncType, role, playbook string, watc
+	validRole := filepath.Join(cwd, "testdata", "roles", "role")
 	testCases := []struct {
-		name  string
-		watch watches.Watch
```
Why it was removed? Are the watches still working well after this change?

djzager replied:
I'm not certain I follow. The structure of the testCase struct was modified so that the watch could be constructed as it is [here](https://github.com/operator-framework/operator-sdk/pull/1852/files/532ebb705ce8856efc05ae64daba426c0a2d0ad8#diff-7160af983c29b9766f0f3db5e9bedd8cR130)

camilamacedo86 commented on pkg/ansible/runner/runner_test.go:
```diff
@@ -63,111 +51,88 @@ func checkCmdFunc(t *testing.T, cmdFunc cmdFuncType, role, playbook string, watc
+		gvk       schema.GroupVersionKind
+		playbook  string
+		role      string
+		finalizer *watches.Finalizer
```
Why is required change the watches and add the finalizer to add support to users defined the verbose level of the ansible container?

djzager replied:
Since I am creating the `watch` below, I need all of the things that could be meaningfully define on a watch to be a part of each test case.

camilamacedo86 commented on pkg/ansible/runner/runner_test.go:
```diff
@@ -177,45 +142,63 @@ func TestNew(t *testing.T) {
+	for _, tc := range testCases {
+		gotString := ansibleVerbosityString(tc.verbosity)
+		if tc.expectedString != gotString {
+			t.Fatalf("Unexpected string %v expected %v", gotString, tc.expectedString)
```
Should not possible to add a test to ensure that if the -vv, for example, be used then the logs are with this verbose? Could we not check that the output which jut will occur in this scenario was not printed?

djzager replied:
I think it would be possible to add a functional test to do this, yes. I do not believe it would be easy to construct a test in `runner_test.go` that does as you describe.

camilamacedo86 replied:
Wdyt about using the e2e tests for it?

djzager commented:
> I looked better this PR and is not clear for me why was required to do some changes?

Part of the "additional" changes are from the fact that this is a continuation of my previous PRs to refactor the Ansible Operator:

1. https://github.com/operator-framework/operator-sdk/pull/1523
1. https://github.com/operator-framework/operator-sdk/pull/1672

That is an important point because, you are correct, more than just making it possible to control Ansible's verbosity is happening here. However, I would argue that as it currently stands where values are coming from is not straightforward...some are being handled in the `ansible` package while some others are being handled in the `runner` package or are hardcoded. 

What this pull request aims to do, in addition to the verbosity stuff, was to centralize:

1. Where the logic for how we determine maxWorkers/ansibleVerbosity is located. It's now in the `watches` package.
1. On what these values are stored, the `watch` struct.

> Also, it is missing add the info over it in the [doc](https://github.com/operator-framework/operator-sdk/blob/master/doc/ansible/user-guide.md#view-the-ansible-logs)

This I need to update. Should I leave the discussion of using `ANSIBLE_VERBOSITY` since I see no reason why that wouldn't work...or should I only discuss `ansible-verbosity` flag + GVK specific environment variables?

camilamacedo86 commented:
HI @djzager, 

IMHO this PR has more than one purpose which makes hard to track these changes in the changelog, review and etc ... since it is not only about to add the flag as described there.  Also, it has many changes in the tests as well which make harder we ensure that all still working properly. So, IMHO you should have a PR for each purpose instead of merging this one in order to help us rewview, ensure the quality, test and etc ... However, I will leave the decision to the team.

Hi @fabianvf, 
Since I think you are who has better knowledge over it, could you help us by reviewing this PR?
Please, feel free to approve and etc if you think that it is ok to merge.

camilamacedo86 commented:
/test sanity

djzager commented:
> IMHO this PR has more than one purpose which makes hard to track these changes in the changelog, review and etc ... since it is not only about to add the flag as described there. Also, it has many changes in the tests as well which make harder we ensure that all still working properly. So, IMHO you should have a PR for each purpose instead of merging this one in order to help us rewview, ensure the quality, test and etc ... However, I will leave the decision to the team.

I do not disagree that there is more going on here than just "make ansible verbosity configurable". I do not particularly _want_ to break this PR up into the 3-5 pieces but **I will** if that's what you need from me to review this PR.

I will wait on feedback from @camilamacedo86, @fabianvf, @jmrodri before doing any more work on this PR (specifically fixing the broken sanity test and addressing some of the remaining comments).

camilamacedo86 commented:
Hi @djzager,

Really thank you for comprehension and I think is a good approach wait for the others review. Maybe they have more knowledge on this part and are ok with move in this way.

djzager commented:
Without a good reason not to break up this work I have decided to go ahead and break this PR up into it's proper pieces.

camilamacedo86 commented:
Really thank you @djzager for your contribution and understanding in this matter.
